### PR TITLE
add container probe duration metrics

### LINKED ISSUE #101035: Create a duration metric for probes
/kind feature
/sig node

As exec probe timeout [bug was fixed](https://github.com/kubernetes/enhancements/issues/1972), upgrade of k8s may lead to unexpected workload disruptions if the exec probe timeout is not corrected properly. Unfortunately there is no easy way today to guess which timeout should be set for the probe. The suggestion is to introduce a duration metric to allow customers to measure this time.

Likely here where other metrics are defined: /pkg/kubelet/prober/prober_manager.go

KEP: https://github.com/kubernetes/enhancements/issues/1972

/assign @jackfrancis
/cc @matthyx

### PR DESCRIPTION
**Pull request by jackfrancis**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR adds duration metrics for kubelet container probes.

#### Which issue(s) this PR fixes:

Fixes #101035

#### Does this PR introduce a user-facing change?

```release-note
add container probe duration metrics
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
add container probe duration metrics
```

### CODE REVIEW

**Comment by jackfrancis** [comment]
@SergeyKanzhelev how does this look as a draft change for probe duration metrics? are we on the right track?

**Comment by mmiranda96** [comment]
/triage accepted
/priority important-soon

Could you update the proper unit tests for the new feature? Also, I'm not 100% sure on this, but it looks like the jobs are flaking. Rerunning them...
/retest

**Comment by mmiranda96** [comment]
Also, please edit the section **Does this PR introduce a user-facing change?**, since it does introduce a new change.

**Comment by pacoxu** [comment]
/release-note-edit

```
add container probe duration metrics
```

**Comment by pacoxu** [comment]
/release-note-edit
```release-note
add container probe duration metrics
```

File: pkg/kubelet/prober/prober_manager.go
Comment (dgrisonnet) [inline pkg/kubelet/prober/prober_manager.go:63]: All these labels are very high cardinality and go against our [instrumentation guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md#dimensionality--cardinality). Reason being that for this histogram you'll have the following number of timeseries generated:
```
16(histogram) * M_containers * O_pods * P_namespaces * 3(probe_type) * 3(results)
```
Which turns out to be pretty insane and can very easily cause the monitoring platform to blow up. As such I would advise removing these labels. But that might not make the metric meaningful anymore so is there any other information that you could use to differentiate probes besides their types?

File: pkg/kubelet/prober/worker.go
Comment (dgrisonnet) [inline pkg/kubelet/prober/worker.go:164]: Duration metrics are usually only meaningful when measuring successes since failures will trigger different code paths that are not really useful to measure since the request is bound to fail.
A concrete example is to measure SLOs. In that case, you are interested in the percentage of requests that failed and the percentage of requests that are too slow. You'll have the number of failures with the counter, in your case `prober_probe_total` and the number of slow requests via the histogram `prober_probe_duration`. So you'll never need the latency of the failures.

File: pkg/kubelet/prober/prober_manager.go
Comment (jackfrancis) [reply pkg/kubelet/prober/prober_manager.go:63]: As far as I can tell all of these labels are required to define uniqueness w/ respect to the provenance of the probe. Can "pod_uid" be eliminated?

Any thoughts here @SergeyKanzhelev ?

File: pkg/kubelet/prober/worker.go
Comment (jackfrancis) [reply pkg/kubelet/prober/worker.go:164]: Thanks! will remove

File: pkg/kubelet/prober/prober_manager.go
Comment (dgrisonnet) [reply pkg/kubelet/prober/prober_manager.go:64]: Having both `pod` and `pod_uid` feels a bit redundant, but it does not add more cardinality to the metric since `pod_uid` already has a unique value per pod. So eliminating the `pod_uid` label will not reduce the current cardinality of the metric.

Comment (ehashman) [reply pkg/kubelet/prober/prober_manager.go:64]: There's no real way to reduce the cardinality of this metric while it still being useful IMO, but OTOH I think adding this by default is going to add a ton of load since it's 16 * 9 * # containers.

@dgrisonnet do you think a summary metric would scope that down enough? That at least brings us down to 3ish buckets instead of 16.

/hold

Comment (dgrisonnet) [reply pkg/kubelet/prober/prober_manager.go:64]: I don't think that would help here, my main concern is that this metric has a cardinality of O(container) which can be problematic. Also, compared to the other container metrics exposed by the kubelet that also have a cardinality of O(container), this one will always expose information about containers that have been deleted. From the ingestion side, this is quite a huge difference.

I haven't really looked at the implementation details as I was not familiar with the codebase, but the cardinality should actually be O(probe) here, and I might not be as big of a problem as O(container)? My main concern was that O(container) can lead to cardinality explosion when they are a lot of short-lived containers created in a cluster, e.g CI jobs. But thinking about it again, I don't think it is relevant to add probes to this kind of container, so we might not see the theoretical cardinality spike that comes from O(container).

If we are sure that there are no use cases where short-lived containers use probes, I am fine with adding this metric. Otherwise, we might want to have it gated behind a flag and disabled by default.

Comment (jackfrancis) [reply pkg/kubelet/prober/prober_manager.go:64]: Thanks for the conversation here, folks. If anyone does have domain expertise in terms of the actual code, maybe a remark about what I'm doing wrong here? Test signal definitely suggests these changes are non-functional at this time.

PS I'll rebase now.

File: pkg/kubelet/prober/prober_manager.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/prober/prober_manager.go:56]: I know that today the duration is in seconds, but we started discussions about sub-second duration. (for example: https://github.com/mikebrow/kubernetes/commit/8926bd2c021a63f365d1a89ed465bd2c5abe599b) Can we change this to milliseconds?

Comment (dgrisonnet) [reply pkg/kubelet/prober/prober_manager.go:56]: The unit behind the histogram can be anything, but usually, for duration, seconds are preferred. It is then up to you to define buckets that better suit your needs. The default ones are:
```
.005, .01, .025, .05, .075, .1, .25, .5, .75, 1, 2.5, 5, 7.5, 10
```
So you'll be able to record any operation with duration down to 5 milliseconds.

File: pkg/kubelet/prober/prober_manager.go
Comment (dgrisonnet) [inline pkg/kubelet/prober/prober_manager.go:55]: Histogram metrics should be suffixed by the unit of the metric. In this case, it should be:
```suggestion
		Name:           "probe_duration_seconds",
```

File: pkg/kubelet/prober/prober_manager.go
Comment (dgrisonnet) [reply pkg/kubelet/prober/prober_manager.go:64]: As I mentioned in https://github.com/kubernetes/kubernetes/pull/104484#discussion_r718492374, removing `pod_uid` shouldn't have any impact on the cardinality since it is O(container) or even O(probe) depending on the implementation.

@ehashman correct me if I am wrong but I don't think there are any. We had a cardinality explosion in 1.22 where a change in a metric resulted in exposing 1.6m timeseries.

Comment (SergeyKanzhelev) [reply pkg/kubelet/prober/prober_manager.go:64]: it should help in situation of many short lived pods like was mentioned here: https://github.com/kubernetes/kubernetes/pull/104484#discussion_r743544075 In that case one pod will have a lot of pod_uid's.

Comment (dgrisonnet) [reply pkg/kubelet/prober/prober_manager.go:64]: It will only barely help IMHO. Pods with variable names such as the ones created by replicasets will still be an issue. In this particular case, there should be as many `pod` as `pod_uid` and I don't think pods containing short-lived containers are necessarily stateful.

I would be fine with the following labels: "probe_type", "result", "container", "namespace". But I am still on the fence with `pod`. I can understand that it is an essential information here, but it is concerning in terms of cardinality.

A separate comment regarding `result`, I usually advise against adding error codes to durations metrics since in most cases you don't really care about the latency of requests based on the code that is returned. When setting up SLOs, you most likely want to know the percentage of requests that are failing, you'll get that with the counter metric and you want to know that the overall latency of your requests is under a certain threshold.

File: pkg/kubelet/prober/prober_manager.go
Comment (SergeyKanzhelev) [reply pkg/kubelet/prober/prober_manager.go:56]: these buckets looks OK to me.

File: pkg/kubelet/prober/prober_manager.go
Comment (jackfrancis) [reply pkg/kubelet/prober/prober_manager.go:63]: removed "result" label

Comment (jackfrancis) [reply pkg/kubelet/prober/prober_manager.go:63]: Wow removed "result" and all the E2E tests pass! Can someone explain that to me?

Comment (SergeyKanzhelev) [reply pkg/kubelet/prober/prober_manager.go:63]: maybe other PRs between your tests fixed them. Serial for example is now passing

Comment (jackfrancis) [reply pkg/kubelet/prober/prober_manager.go:63]: 😬

Comment (jackfrancis) [reply pkg/kubelet/prober/prober_manager.go:63]: I'm going to rebase now just so this has the latest bits when it is considered for merge

Comment (dgrisonnet) [reply pkg/kubelet/prober/prober_manager.go:63]: > We already have the probe_total metric with these dimensions. What if we will have twice less buckets? (like 0.25, 0.5, 1, 2, 5, 10, 30, 60 or even 0.5, 1, 5, 10, 30)?

The buckets themselves aren't really the issue here since it basically adds a constant factor to the cardinality. Even though the `probe_total` metric has the same dimensions today, I don't think they are safe at all.

I am very much on the fence here since I can see the value of this metric and each one of its labels, but at the same time, the cardinality of O(container w/ probe) is quite frightening. 

I can see a few options moving forward:
- Go with the assumption that short-lived containers will not have probes. (It might bite us at some point)
- Introduce a mechanism to disable these high-cardinality metrics by default - doesn't exist yet
- Introduce a mechanism to bound the cardinality of these metrics - doesn't exist yet 

@ehashman @logicalhan - any thoughts on that? We could also bring that subject to the next sig meeting

Comment (SergeyKanzhelev) [reply pkg/kubelet/prober/prober_manager.go:63]: > Introduce a mechanism to bound the cardinality of these metrics - doesn't exist yet

Having pod names without deployment and replicaset randomness added for metrics would be great. Maybe also rely on some label if set to override the value of a pod dimension as a feature in the new k8s version? Including this metric today and implementing the name override feature in the next milestone may be a good idea.

> @ehashman @logicalhan - any thoughts on that? We could also bring that subject to the next sig meeting

@ehashman @logicalhan is there any opinion you can share today to include the metric as is into the 1.23 milestone?

Comment (jackfrancis) [reply pkg/kubelet/prober/prober_manager.go:63]: > Introduce a mechanism to disable these high-cardinality metrics by default - doesn't exist yet

I think moving this behind a feature gate makes plenty of sense to me, happy to do that. (If that requires a KEP happy to do so as well.)

**Comment by dgrisonnet** [comment]
As discussed on slack, I am fine with this new metric from a SIG instrumentation perspective, as long as we have a follow-up for the next release reducing the cardinality of the pod label: https://kubernetes.slack.com/archives/C20HH14P7/p1637080128045600

/unhold

File: pkg/kubelet/prober/prober_manager.go
Comment (dgrisonnet) [inline pkg/kubelet/prober/prober_manager.go:63]: `pod_uid` isn't meaningful for the use-case of this metric since we already have the pod name. As such I would be in favor of removing it.

File: pkg/kubelet/prober/prober_manager.go
Comment (ehashman) [reply pkg/kubelet/prober/prober_manager.go:63]: I'm torn on this. We have a mechanism to [disable high-cardinality metrics](https://kubernetes.io/docs/concepts/cluster-administration/system-metrics/#disabling-metrics), but it's opt out as opposed to opt in. A feature gate would be better from that perspective since it's opt-in as opposed to opt-out, but would require a KEP, and that would require a full enhancement cycle which seems like overkill for this.

File: pkg/kubelet/prober/prober_manager.go
Comment (dgrisonnet) [inline pkg/kubelet/prober/prober_manager.go:56]: nit: 
```suggestion
		Help:           "Duration in seconds for a probe response.",
```

File: pkg/kubelet/prober/prober_manager.go
Comment (logicalhan) [reply pkg/kubelet/prober/prober_manager.go]: yeah same concerns as @ehashman on my side...

**Review by logicalhan — COMMENTED** [review]
/lgtm
/approve

I'll put aside my cardinality concerns due to the existence of a kubelet metric which already uses these labels.

**Comment by SergeyKanzhelev** [comment]
failures might be related. Something failing on probes initialization. Can it be the fact that we removed dimensions, would it break things when we copy dimensions?

**Comment by jackfrancis** [comment]
@SergeyKanzhelev I assume you mean the removal of pod_uid?

**Comment by jackfrancis** [comment]
@SergeyKanzhelev re-added back `pod_uid` to attempt to sort of prove that the removal of it is the cause of test failures

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Comment by dgrisonnet** [comment]
Since @ehashman and @logicalhan also expressed some concerns with the current metric, could we perhaps resolve them before proceeding? https://github.com/kubernetes/kubernetes/pull/104484/files#r750568110

From the slack discussion we had before the feature freeze (https://kubernetes.slack.com/archives/C20HH14P7/p1637080128045600), the most reasonable solution seems to be to remove the `pod_uid` label and prune the `uid` from the pod name before adding it to the `pod` label. For consistency, I think we should also apply these to the `probe_total` metric.

Bringing the original hold back, until we have addressed these points.
/hold

**Comment by jackfrancis** [comment]
@dgrisonnet removing the `pod_uid` label produced broken test case outcomes. Not saying the concerns aren't valid, just FYI we haven't proven that removing the `pod_uid` is a functional change.

**Comment by dgrisonnet** [comment]
Since the metric is introduced by this PR, I wouldn't expect existing tests to depend on its `pod_uid` label, so it's weird to me that this could break test cases. Were you able to verify that the tests we failing because of this change or could it have been flakes?

**Comment by dgrisonnet** [comment]
Yes, please. We can add this label to the metric since it would cause cardinality issues with ephemeral pods.

**Comment by jackfrancis** [comment]
@dgrisonnet done with this commit: https://github.com/kubernetes/kubernetes/pull/104484/commits/344ed01a613c3a2da3f9bd80bfaaef5bffc28810

Current E2E jobs will tell the tale

**Comment by dgrisonnet** [comment]
Even though you removed the `pod_uid` label from the metric definition, you are still providing it when updating the metric which I suppose result in a fatal error.

**Comment by jackfrancis** [comment]
@dgrisonnet good catch! fixed

**Comment by jackfrancis** [comment]
@dgrisonnet @SergeyKanzhelev PTAL

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

**Comment by jackfrancis** [comment]
Thx @dims 

@dgrisonnet @SergeyKanzhelev do we want to consider merging this without the `pod_uid` label?

**Comment by dgrisonnet** [comment]
This comment still applies https://github.com/kubernetes/kubernetes/pull/104484#issuecomment-989902194, we should remove the uid part of the pod name before proceeding with this PR.

File: pkg/kubelet/prober/worker.go
Comment (jackfrancis) [inline pkg/kubelet/prober/worker.go:115]: @dgrisonnet is this what we had in mind?

I can clean up the implementation after we agree on an approach (is there a common lib that will do this pod UID redaction in a more resilient way?), write UT, etc.

cc @SergeyKanzhelev

Comment (dgrisonnet) [reply pkg/kubelet/prober/worker.go:115]: Yeah, that's similar to what I had in mind but I don't think that will cover all the use-cases. We have a few possible scenarios here:
1. static pods / pods created manually
2. pods created by a replicaset/statefulset
3. pods created by a deployment

Scenario 2 can be handled by relying on pod.GenerateName which contains pod.Name + "-".

For scenario 3 it is a bit more complex because the resulting pod will be named pod.Name + pod template hash + uid and pod.GenerateName will be equal to pod.Name + "-" + pod template hash + "-". We can rely on pod.Labels["pod-template-hash"] to get the pod template hash and then deduce the pod name without UIDs.

As for scenario 1, we don't want to do anything since they won't have UIDs. We just have to make sure that it is still possible to have names with hyphens and without UIDs.

So my suggestion would to have something like:
```go
podName = w.pod.Name
if w.pod.GenerateName != "" {
  podNameSlice := strings.Split(w.pod.Name, "-")
  podName = strings.Join(podNameSlice[:len(podNameSlice)-1], "-")
  if label, ok := w.pod.GetLabels()[apps.DefaultDeploymentUniqueLabelKey]; ok {
    podName = strings.ReplaceAll(podName, fmt.Sprintf("-%s", label), "")
  }
}
```
It is pretty rough, so not sure if that would be the best option to go with, but it should give you a better idea of the different scenarios that we need to support here.

File: pkg/kubelet/prober/worker_test.go
Comment (jackfrancis) [inline pkg/kubelet/prober/worker_test.go:479]: @dgrisonnet thx for the continual feedback!

Does the below set of cases prove out everything we want?

Comment (jackfrancis) [reply pkg/kubelet/prober/worker_test.go:479]: gentle ping @dgrisonnet @SergeyKanzhelev 

This PR is ready for another "consider for merge" review

Comment (dgrisonnet) [reply pkg/kubelet/prober/worker_test.go:479]: The new changes looks good from my side, but I would love to have a second pair of eyes on these assumptions: https://github.com/kubernetes/kubernetes/pull/104484#discussion_r788750303

File: pkg/kubelet/prober/worker.go
Comment (jackfrancis) [reply pkg/kubelet/prober/worker.go:298]: is this comment still valid? sorry if I haven't addressed it, I don't quite understand what you mean

**Comment by jackfrancis** [comment]
gentle ping @pacoxu

thank you!

**Comment by pacoxu** [comment]
I did some testing. Here are four examples including static pod and general pods.

```
[root@daocloud ~]# kubectl get pod --show-labels -A
NAMESPACE          NAME                                       READY   STATUS             RESTARTS         AGE     LABELS
default            liveness-exec                              0/1     CrashLoopBackOff   23 (54s ago)     75m     test=liveness
calico-system      calico-kube-controllers-59c45ff85c-x2gts   1/1     Running            10 (27d ago)     56d     k8s-app=calico-kube-controllers,pod-template-hash=59c45ff85c
calico-system      calico-node-d2jpf                          1/1     Running            4 (27d ago)      56d     controller-revision-hash=86b6bcc6dc,k8s-app=calico-node,pod-template-generation=1
calico-system      calico-typha-65cdbb598d-644x5              1/1     Running            7 (27d ago)      56d     k8s-app=calico-typha,pod-template-hash=65cdbb598d
kube-system        kube-apiserver-daocloud                    1/1     Running            0                81m     component=kube-apiserver,tier=control-plane
```
The metrics for pods above 
```
prober_probe_duration_seconds_sum{container="liveness",namespace="default",pod="liveness-exec",probe_type="Liveness"} 6.563395632
prober_probe_duration_seconds_sum{container="calico-kube-controllers",namespace="calico-system",pod="calico-kube-controllers",probe_type="Liveness"} 25.634286655000015
prober_probe_duration_seconds_sum{container="calico-kube-controllers",namespace="calico-system",pod="calico-kube-controllers",probe_type="Readiness"} 26.072259959999986
prober_probe_duration_seconds_sum{container="calico-node",namespace="calico-system",pod="calico-node",probe_type="Liveness"} 0.6949132140000004
prober_probe_duration_seconds_sum{container="calico-node",namespace="calico-system",pod="calico-node",probe_type="Readiness"} 60.58489301800003
prober_probe_duration_seconds_sum{container="calico-typha",namespace="calico-system",pod="calico-typha",probe_type="Liveness"} 0.7015105350000005
prober_probe_duration_seconds_sum{container="calico-typha",namespace="calico-system",pod="calico-typha",probe_type="Readiness"} 0.6845843959999995
prober_probe_duration_seconds_sum{container="kube-apiserver",namespace="kube-system",pod="kube-apiserver-daocloud",probe_type="Liveness"} 3.810594282999996
prober_probe_duration_seconds_sum{container="kube-apiserver",namespace="kube-system",pod="kube-apiserver-daocloud",probe_type="Readiness"} 44.2331933840001
prober_probe_duration_seconds_sum{container="kube-apiserver",namespace="kube-system",pod="kube-apiserver-daocloud",probe_type="Startup"} 0.006615101
```

**Review by pacoxu — COMMENTED** [review]
/lgtm

File: pkg/kubelet/prober/worker.go
Comment (pacoxu) [reply pkg/kubelet/prober/worker.go]: @dgrisonnet 
can we unhold this pr now?

**Comment by jackfrancis** [comment]
@SergeyKanzhelev can we merge this?

**Comment by jackfrancis** [comment]
@SergeyKanzhelev @pacoxu @dgrisonnet should we take advantage of the beginning of the 1.25 cycle and add this to the milestone and merge?

**Comment by pacoxu** [comment]
kindly ping @dgrisonnet 
/cc  @dashpole @mrunalp
Do you think this metric is acceptable?

**Comment by dgrisonnet** [comment]
I somehow missed the pings, sorry about that. I think this PR is good to go now that my initial concerns have been addressed.

/unhold
/lgtm

**Comment by jackfrancis** [comment]
gentle ping @mrunalp

**Review by mrunalp — APPROVED** [review]

**Pull request merged**
